tiledb/sm/query/reader.cc

```diff
@@ -3200,30 +3213,175 @@ Status Reader::copy_coordinates(
+            rcs_partition,
+            partition_start,
+            partition_end,
+            *result_cell_slabs,
+            stride);
+      });
+
+  for (const auto& st : statuses)
+    RETURN_NOT_OK(st);
+
+  // Calculate the total number of result cell slabs.
+  uint64_t num_rcs = 0;
+  for (const auto& rcs_partition : rcs_partitions) {
+    num_rcs += rcs_partition.size();
+  }
+
+  // Concatenate all of the result cell slabs in the partitions
+  // into `result_cell_slabs`.
+  *result_cell_slabs = std::move(rcs_partitions[0]);
+  result_cell_slabs->reserve(num_rcs);
+  for (size_t p = 1; p < rcs_partitions.size(); ++p) {
+    std::vector<ResultCellSlab>* const rcs_partition = &rcs_partitions[p];
+    result_cell_slabs->insert(
+        result_cell_slabs->end(),
+        std::make_move_iterator(rcs_partition->begin()),
+        std::make_move_iterator(rcs_partition->end()));
+  }
+
+  return Status::Ok();
+}
+
+Status Reader::apply_query_condition_partition(
+    std::vector<ResultCellSlab>* const rcs_partition,
+    const size_t partition_start,
+    const size_t partition_end,
+    const std::vector<ResultCellSlab>& result_cell_slabs,
+    const uint64_t stride) {
+  // TODO handle tile == nullptr (empty range)
+
+  for (size_t r = partition_start; r < partition_end; ++r) {
+    const ResultCellSlab& rcs = result_cell_slabs.at(r);
+    ResultTile* const tile = rcs.tile_;
+    const uint64_t start = rcs.start_;
+    const uint64_t length = rcs.length_;
+
+    // Handle an empty range.
+    if (tile == nullptr) {
+      bool cmp;
+      condition_.cmp_cell_fill_value(array_schema_, &cmp);
+      if (cmp) {
+        rcs_partition->emplace_back(tile, start, length);
+      }
+    } else {
+      // Start the pending result cell slab at the start position
+      // of the current result cell slab.
+      uint64_t pending_start = start;
+
+      // Iterate through each cell in this slab.
+      uint64_t c;
+      for (c = 0; c < length; ++c) {
+        bool cmp;
+        RETURN_NOT_OK(condition_.cmp_cell(
+            array_schema_, tile, start + (stride * c), &cmp));
+
+        if (!cmp) {
+          // Create a result cell slab if there are pending cells.
+          if (pending_start != start + c) {
+            const uint64_t rcs_start =
+                start + ((pending_start - start) * stride);
+            const uint64_t rcs_length = c - (pending_start - start);
+            rcs_partition->emplace_back(tile, rcs_start, rcs_length);
+          }
+
+          // Set the start of the pending result cell slab.
+          pending_start = start + c + 1;
+        }
+      }
+
+      // Create the final result cell slab if there are pending cells.
+      if (pending_start != start + c) {
+        const uint64_t rcs_start = start + ((pending_start - start) * stride);
+        const uint64_t rcs_length = c - (pending_start - start);
+        rcs_partition->emplace_back(tile, rcs_start, rcs_length);
```

Comment:
Also this is potentially a performance bottleneck due to the potentially large number of new object creation. Perhaps a bytemap would work better here over the initial result cell slabs.